docs/en/sql-reference/functions/string-search-functions.md

```diff
@@ -536,4 +536,44 @@ For case-insensitive search or/and in UTF-8 format use functions `ngramSearchCas
 !!! note "Note"
     For UTF-8 case we use 3-gram distance. All these are not perfectly fair n-gram distances. We use 2-byte hashes to hash n-grams and then calculate the (non-)symmetric difference between these hash tables – collisions may occur. With UTF-8 case-insensitive format we do not use fair `tolower` function – we zero the 5-th bit (starting from zero) of each codepoint byte and first bit of zeroth byte if bytes more than one – this works for Latin and mostly for all Cyrillic letters.
 
+## substringCount(haystack, needle) {#substringCount}
+
+Count the number of substring occurrences
+
+For a case-insensitive search, use the function `substringCountCaseInsensitive`.
+
+**Syntax**
+
+``` sql
+substringCount(haystack, needle[, start_pos])
+```
+
+**Parameters**
+
+-   `haystack` — The string to search in. [String](../../sql-reference/syntax.md#syntax-string-literal).
+-   `needle` — The substring to search for. [String](../../sql-reference/syntax.md#syntax-string-literal).
+-   `start_pos` – Optional parameter, position of the first character in the string to start search. [UInt](../../sql-reference/data-types/int-uint.md)
+
+**Returned values**
+
+-   Number of occurrences.
```

Comment:
How to count if substring is empty?

(throw an exception; return the number of chars in haystack; return the number of chars + 1 - all three variants Ok).

Reply:
No mention about intersecting substrings in haystack.
Example: 

```
substringCount('aaaa', 'aa')
```

may return 2 for non-intersecting substrings:
```
aaaa
aa
  aa
```

or 3 for intersecting substrings:
```
aaaa
aa
 aa
  aa
```

Reply:
(it's Ok to return non-intersecting substrings only)

Reply:
Let's also add a test to check this behaviour.

Reply:
It's better to count non-intersecting substrings, because it will be consistent with the `extractAll` function.

Reply:
The number of returned substrings for empty needle should be consistent with the returned array size of `extractAll` when empty regexp is passed.

Reply:
>How to count if substring is empty?

It will return zero (same as `extractAll`, tests updated)

>The number of returned substrings for empty needle should be consistent with the returned array size of extractAll when empty regexp is passed.

Yes, it will also return zero (tests updated)

>It's better to count non-intersecting substrings, because it will be consistent with the extractAll function.

In the latest version this has been addressed (tests updated for all variants and add some comments in code just in case)